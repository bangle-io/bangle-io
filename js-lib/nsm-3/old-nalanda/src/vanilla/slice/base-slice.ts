import type { EffectStore } from '../effect';
import type { UpdaterType } from '../helpers';
import { idGeneration, Updater } from '../helpers';
import type { StoreState } from '../store-state';
import type { SliceId } from '../types';
import type { Slice } from './slice';

export type UserSliceOpts<
  TSliceName extends string,
  TState extends object,
  TDep extends string,
> = {
  name: TSliceName;
  state: TState;
  dependencies: Array<Slice<TDep, any, any>>;
};

export type CalcDerivedState = (
  storeState: StoreState<any>,
) => Record<string, unknown>;

export type CreateSliceOpts<
  TSliceName extends string,
  TState extends object,
  TDep extends string,
> = UserSliceOpts<TSliceName, TState, TDep> & {
  // provided internally
  sliceId?: SliceId;
  calcDerivedState?: CalcDerivedState;
};

export type ValidStoreState<
  TStoreSlices extends string,
  TSliceName extends string,
> = TSliceName extends TStoreSlices ? StoreState<TStoreSlices> : never;

export type ValidEffectStore<
  TStoreSlices extends string,
  TSliceName extends string,
> = TSliceName extends TStoreSlices ? EffectStore<TStoreSlices> : never;

export abstract class BaseSlice<
  TSliceName extends string,
  TState extends object,
  TDep extends string,
> {
  readonly name: TSliceName;
  readonly initialState: TState;
  readonly dependencies: Array<Slice<TDep, any, any>>;
  readonly sliceId: SliceId;

  constructor(opts: CreateSliceOpts<TSliceName, TState, TDep>) {
    this.name = opts.name;
    this.initialState = opts.state;
    this.dependencies = opts.dependencies;
    this.sliceId = opts.sliceId || idGeneration.createSliceId(opts.name);
  }

  get<TStoreSlices extends string>(
    storeState: ValidStoreState<TStoreSlices, TSliceName>,
  ): TState {
    return storeState.resolve(this.sliceId) as TState;
  }

  query<TParams extends any[], TQuery>(
    cb: (...params: TParams) => (storeState: StoreState<TSliceName>) => TQuery,
  ): // TODO - currently TDep is not included as part of storeState
  (storeState: StoreState<any>, ...params: TParams) => TQuery {
    return (storeState, ...params) => cb(...params)(storeState);
  }

  update<TStoreSlices extends string>(
    storeState: ValidStoreState<TStoreSlices, TSliceName>,
    updater: ((cur: TState) => Partial<TState>) | Partial<TState>,
    opts: { replace?: boolean } = {},
  ): UpdaterType<TSliceName> {
    const sliceState = storeState.getSliceStateManager(this.sliceId)
      .sliceState as TState;

    const newSliceState =
      typeof updater === 'function' ? updater(sliceState) : updater;

    let mergedState: Partial<TState>;

    if (opts.replace) {
      mergedState = newSliceState;
    }
    // maintain reference equality if possible
    else if (newSliceState === sliceState) {
      mergedState = newSliceState;
    } else {
      mergedState = { ...sliceState, ...newSliceState };
    }

    return {
      name: this.name,
      [Updater]: mergedState,
    };
  }
}
