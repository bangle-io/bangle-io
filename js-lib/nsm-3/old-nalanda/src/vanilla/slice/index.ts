export type { ValidStoreState } from './base-slice';
export { BaseSlice } from './base-slice';
export { Slice, slice } from './slice';
export type { SliceKey } from './slice-key';
export { sliceKey } from './slice-key';
