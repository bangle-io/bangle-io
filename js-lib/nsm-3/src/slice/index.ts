export { BaseSlice } from './base-slice';
export { Slice, slice } from './slice';
export { sliceKey } from './slice-key';
