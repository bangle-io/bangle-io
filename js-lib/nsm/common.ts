import type { ActionSerializer } from './action-serializer';
import type { Slice } from './slice';
import type { StoreState } from './state';
import type { Store } from './store';

export type AnyFn = (...args: any[]) => any;

export const expectType = <Type>(_: Type): void => void 0;

export type ReplaceReturnType<T extends (...args: any) => any, R> = T extends (
  ...args: infer P
) => any
  ? (...ag: P) => R
  : never;

export type SelectorFn<SS, DS extends Slice[], T> = (
  sliceState: SS,
  storeState: StoreState<DS>,
) => T;

export type RawAction<P extends any[], SS, DS extends Slice[]> = (
  ...payload: P
) => (sliceState: SS, storeState: StoreState<DS>) => SS;

export type Action<K extends string, P extends any[] = unknown[]> = (
  ...payload: P
) => Transaction<K, P>;

export type InferSliceKey<SL extends AnySliceBase> = SL extends SliceBase<
  infer K,
  any
>
  ? K
  : never;

export type InferSliceDep<SL extends AnySliceBase> = SL extends Slice<
  any,
  any,
  infer DS
>
  ? DS[number]
  : never;

// returns a union of all slice keys
export type InferSlicesKey<SlicesRegistry extends AnySliceBase[]> = {
  [K in keyof SlicesRegistry]: InferSliceKey<SlicesRegistry[K]>;
}[number];

// returns a slice if it is registered in the slice registry
export type ResolveSliceIfRegistered<
  SL extends AnySliceBase,
  SliceRegistry extends AnySliceBase[],
> = SL extends SliceBase<infer K, any>
  ? K extends InferSlicesKey<SliceRegistry>
    ? SL
    : never
  : never;

export interface SliceKeyBase<K extends string, SS> {
  key: K;
  initState: SS;
  dependencies: Array<SliceBase<string, unknown>>;
}

export interface EffectsBase<SL extends Slice = Slice> {
  updateSync?: (
    sl: SL,
    store: Store<Array<SL | InferSliceDep<SL>>>,
    prevStoreState: StoreState<Array<SL | InferSliceDep<SL>>>,
  ) => void;
  update?: (
    sl: SL,
    store: Store<Array<SL | InferSliceDep<SL>>>,
    prevStoreState: StoreState<Array<SL | InferSliceDep<SL>>>,
  ) => void;
  once?: (
    sl: SL,
    store: Store<Array<SL | InferSliceDep<SL>>>,
  ) => void | (() => void);
}

export type AnySliceBase = SliceBase<string, unknown>;

export interface SliceBase<K extends string, SS> {
  key: SliceKeyBase<K, SS>;
  fingerPrint: string;
  effects?: Array<EffectsBase<any>>;
  // TODO make any to SS
  _actionSerializer: ActionSerializer<K>;
}

export interface StoreTransaction<K extends string, P extends unknown[]>
  extends Transaction<K, P> {
  id: string;
}

export interface Transaction<K extends string, P extends unknown[]> {
  sliceKey: K;
  actionId: string;
  payload: P;
}

export function mapObjectValues<T, U>(
  obj: Record<string, T>,
  fn: (v: T, k: string) => U,
): Record<string, U> {
  const newObj: Record<string, U> = Object.create(null);

  for (const [key, value] of Object.entries(obj)) {
    newObj[key] = fn(value, key);
  }

  return newObj;
}

export function objectHasOwnProperty<X extends {}, Y extends PropertyKey>(
  obj: X,
  prop: Y,
): obj is X & Record<Y, unknown> {
  return Object.prototype.hasOwnProperty.call(obj, prop);
}

export function calcReverseDependencies(
  slices: AnySliceBase[],
): Record<string, Set<string>> {
  let reverseDependencies: Record<string, Set<string>> = {};

  for (const slice of slices) {
    for (const dep of slice.key.dependencies) {
      let result = reverseDependencies[dep.key.key];

      if (!result) {
        result = new Set();
        reverseDependencies[dep.key.key] = result;
      }

      result.add(slice.key.key);
    }
  }

  return reverseDependencies;
}

export function flattenReverseDependencies(
  reverseDep: Record<string, Set<string>>,
) {
  const result: Record<string, Set<string>> = {};

  const recurse = (key: string) => {
    let vals = result[key];

    if (vals) {
      return vals;
    }

    vals = new Set<string>();
    result[key] = vals;

    const deps = reverseDep[key];

    if (deps) {
      for (const dep of deps) {
        vals.add(dep);
        for (const v of recurse(dep)) {
          vals.add(v);
        }
      }
    }

    return vals;
  };

  for (const key of Object.keys(reverseDep)) {
    recurse(key);
  }

  return result;
}

export function calcDependencies(
  slices: AnySliceBase[],
): Record<string, Set<string>> {
  return Object.fromEntries(
    slices.map((slice) => [
      slice.key.key,
      new Set(slice.key.dependencies.map((dep) => dep.key.key)),
    ]),
  );
}
