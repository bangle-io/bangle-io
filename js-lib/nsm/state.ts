import type {
  AnySliceBase,
  ResolveSliceIfRegistered,
  Transaction,
} from './types';

interface StoreStateOptions {
  debug?: boolean;
}

export interface StoreStateConfig<SB extends AnySliceBase> {
  slices: SB[];
  opts?: StoreStateOptions;
}
export class StoreState<SB extends AnySliceBase = any> {
  static checkDependencyOrder(slices: AnySliceBase[]) {
    let seenKeys = new Set<string>();
    for (const slice of slices) {
      const { key } = slice;

      if (key.dependencies !== undefined) {
        const depKeys = key.dependencies.map((d) => d.key.key);

        for (const depKey of depKeys) {
          if (!seenKeys.has(depKey)) {
            throw new Error(
              `Slice "${key.key}" has a dependency on Slice "${depKey}" which is either not registered or is registered after this slice.`,
            );
          }
        }
      }
      seenKeys.add(key.key);
    }
  }

  static checkUniqueKeys(slices: AnySliceBase[]) {
    const keys = slices.map((s) => s.key.key);
    const unique = new Set(keys);

    if (keys.length !== unique.size) {
      const dups = findDuplications(keys);
      throw new Error('Duplicate slice keys ' + dups.join(', '));
    }
  }

  static create<SB extends AnySliceBase>({
    slices,
    opts,
  }: StoreStateConfig<SB>): StoreState<SB> {
    StoreState.checkUniqueKeys(slices);
    StoreState.checkDependencyOrder(slices);

    const instance = new StoreState(slices, opts);

    for (const slice of slices) {
      instance.slicesCurrentState[slice.key.key] = slice.key.initState;
    }

    return instance;
  }

  protected slicesCurrentState: { [k: string]: any } = Object.create(null);

  protected _transaction: undefined | Transaction<string, unknown[]> =
    undefined;

  constructor(public _slices: SB[], public opts?: StoreStateOptions) {}

  /**
   * Get the transaction that created this state
   */
  get transaction() {
    return this._transaction;
  }

  applyTransaction<P extends any[]>(
    tx: Transaction<SB['key']['key'], P>,
  ): StoreState<SB> {
    const newState = { ...this.slicesCurrentState };

    let found = false;

    for (const slice of this._slices) {
      if (slice.key.key === tx.sliceKey) {
        found = true;
        const rawAction = slice._actionSerializer.getRawAction(tx.actionId);

        if (!rawAction) {
          throw new Error(
            `Action "${tx.actionId}" not found in Slice "${slice.key.key}"`,
          );
        }

        const sliceState = this._getSliceState(slice);

        if (!sliceState.found) {
          throw new Error(
            `Slice "${slice.key.key}" or one of its dependencies not found in store`,
          );
        }

        newState[slice.key.key] = rawAction(...tx.payload)(
          sliceState.value,
          this,
        );
      }
    }

    if (!found) {
      return this;
    }

    // TODO: append-action
    return this._fork(newState, tx);
  }

  getSliceState<SL extends AnySliceBase>(
    slice: ResolveSliceIfRegistered<SL, SB>,
  ): SL['key']['initState'] {
    let result = this.slicesCurrentState[slice.key.key]!;

    if (result === undefined) {
      throw new Error(`Slice "${slice.key.key}" not found in store`);
    }

    return result;
  }

  private _fork(
    slicesState: StoreState['slicesCurrentState'],
    tx: Transaction<string, unknown[]>,
  ): StoreState {
    const newInstance = new StoreState(this._slices, this.opts);

    newInstance._transaction = tx;
    newInstance.slicesCurrentState = slicesState;

    return newInstance;
  }

  // An internal method to get the state of a slice without generic type hassles
  private _getSliceState<SL extends AnySliceBase>(
    slice: SL,
  ):
    | { found: true; value: SL['key']['initState'] }
    | { found: false; value: undefined } {
    if (Object.hasOwnProperty.call(this.slicesCurrentState, slice.key.key)) {
      return {
        found: true,
        value: this.slicesCurrentState[slice.key.key],
      };
    }

    return { found: false, value: undefined };
  }
}

function findDuplications<T>(arr: T[]): T[] {
  const seen = new Set<T>();
  const dupes = new Set<T>();

  for (const item of arr) {
    if (seen.has(item)) {
      dupes.add(item);
    } else {
      seen.add(item);
    }
  }

  return [...dupes];
}
