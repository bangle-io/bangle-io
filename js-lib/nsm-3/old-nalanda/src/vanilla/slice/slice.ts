import type { UserActionCallback } from '../action';
import { Action, ActionBuilder } from '../action';
import type { UpdaterType } from '../helpers';
import type { StoreState } from '../store-state';
import type { Transaction } from '../transaction';
import type { AnySlice, InferSliceNameFromSlice } from '../types';
import type {
  CreateSliceOpts,
  UserSliceOpts,
  ValidEffectStore,
} from './base-slice';
import { BaseSlice } from './base-slice';

export class Slice<
  TSliceName extends string,
  TState extends object,
  TDep extends string,
> extends BaseSlice<TSliceName, TState, TDep> {
  /**
   * @internal
   */
  static create<
    TSliceName extends string,
    TState extends object,
    TDepSlice extends Slice<string, any, any>,
  >(
    opts: CreateSliceOpts<
      TSliceName,
      TState,
      InferSliceNameFromSlice<TDepSlice>
    >,
  ): Slice<TSliceName, TState, InferSliceNameFromSlice<TDepSlice>> {
    return new Slice(opts);
  }

  private constructor(
    public readonly opts: CreateSliceOpts<TSliceName, TState, TDep>,
  ) {
    super(opts);
  }

  simpleAction<TKey extends keyof TState>(
    key: TKey,
    updater?: (
      val: TState[TKey],
      state: StoreState<TSliceName | TDep>,
    ) => Partial<TState>,
  ) {
    return this.action(
      (val: TState[TKey]) => {
        return this.tx((state) => {
          return this.update(state, (existing) => {
            if (!updater) {
              if (val === existing[key]) {
                return existing;
              }

              return {
                ...existing,
                [key]: val,
              };
            }

            const newVal = updater(val, state);

            if (newVal === existing[key]) {
              return existing;
            }

            return newVal;
          });
        });
      },
      {
        name: `simpleAction(${key.toString()})`,
      },
    );
  }

  action<TParams extends any[]>(
    cb: UserActionCallback<TParams, ActionBuilder<any, any>>,
    opts?: { name?: string },
  ): (...params: TParams) => Transaction<TSliceName> {
    const action = new Action<TSliceName, TParams>({
      slice: this,
      userCallback: cb,
      debugName: opts?.name,
    });

    return action.getTransactionBuilder();
  }

  track<TStoreSlices extends string>(
    store: ValidEffectStore<TStoreSlices, TSliceName>,
  ): TState {
    return new Proxy(this.get(store.state as StoreState<any>), {
      get: (target, prop: string) => {
        // @ts-expect-error not sure how to fix this
        const val = target[prop];

        store._addTrackedField(this, prop, val);

        return val;
      },
    }) as any;
  }

  tx(
    calcSliceState: (
      storeState: StoreState<TSliceName | TDep>,
    ) => TState | UpdaterType<TSliceName>,
  ): ActionBuilder<TSliceName, any> {
    return new ActionBuilder({
      name: this.name,
      calcUserSliceState: calcSliceState,
    });
  }
}

export function slice<
  TSliceName extends string,
  TState extends object,
  TDepSlice extends Slice<string, any, any>,
>(
  dependencies: TDepSlice[],
  opts: Omit<
    UserSliceOpts<TSliceName, TState, InferSliceNameFromSlice<TDepSlice>>,
    'dependencies'
  >,
): Slice<TSliceName, TState, InferSliceNameFromSlice<TDepSlice>> {
  return Slice.create<TSliceName, TState, AnySlice>({ ...opts, dependencies });
}
