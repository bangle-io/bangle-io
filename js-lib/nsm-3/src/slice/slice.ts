import type { UserActionCallback } from '../action';
import { Action, ActionBuilder } from '../action';
import type { StoreState } from '../store-state';
import type { Transaction } from '../transaction';
import type { AnySlice, InferSliceNameFromSlice } from '../types';
import type {
  CreateSliceOpts,
  UpdaterType,
  UserSliceOpts,
  ValidEffectStore,
} from './base-slice';
import { BaseSlice } from './base-slice';

export class Slice<
  TSliceName extends string,
  TState extends object,
  TDep extends string,
> extends BaseSlice<TSliceName, TState, TDep> {
  /**
   * @internal
   */
  static create<
    TSliceName extends string,
    TState extends object,
    TDepSlice extends Slice<string, any, any>,
  >(
    opts: CreateSliceOpts<
      TSliceName,
      TState,
      InferSliceNameFromSlice<TDepSlice>
    >,
  ): Slice<TSliceName, TState, InferSliceNameFromSlice<TDepSlice>> {
    return new Slice(opts);
  }

  private constructor(
    public readonly opts: CreateSliceOpts<TSliceName, TState, TDep>,
  ) {
    super(opts);
  }

  action<TParams extends any[]>(
    cb: UserActionCallback<TParams, ActionBuilder<any, any>>,
  ): (...params: TParams) => Transaction<TSliceName> {
    const action = new Action<TSliceName, TParams>({
      slice: this,
      userCallback: cb,
    });

    return action.getTransactionBuilder();
  }

  //   TODO implement
  query<TParams extends any[], TQuery>(
    cb: (
      ...params: TParams
    ) => (storeState: StoreState<TSliceName | TDep>) => TQuery,
  ): (storeState: StoreState<TSliceName | TDep>, ...params: TParams) => TQuery {
    return cb as any;
  }

  track<TStoreSlices extends string>(
    store: ValidEffectStore<TStoreSlices, TSliceName>,
  ): {
    [TKey in keyof TState]-?: () => TState[TKey];
  } {
    const result = Object.fromEntries(
      Object.keys(this.get(store.state as StoreState<any>)).map((key) => [
        key,
        () => {
          return (this.get(store.state as StoreState<any>) as any)[key];
        },
      ]),
    );

    return new Proxy(result, {
      get: (target, prop: string) => {
        store._addTrackedField(this, prop, target[prop]!());

        return target[prop];
      },
    }) as any;
  }

  tx(
    calcSliceState: (
      storeState: StoreState<TSliceName | TDep>,
    ) => TState | UpdaterType<TSliceName>,
  ): ActionBuilder<TSliceName, any> {
    return new ActionBuilder({
      name: this.name,
      calcUserSliceState: calcSliceState,
    });
  }
}

export function slice<
  TSliceName extends string,
  TState extends object,
  TDepSlice extends Slice<string, any, any>,
>(
  dependencies: TDepSlice[],
  opts: Omit<
    UserSliceOpts<TSliceName, TState, InferSliceNameFromSlice<TDepSlice>>,
    'dependencies'
  >,
): Slice<TSliceName, TState, InferSliceNameFromSlice<TDepSlice>> {
  return Slice.create<TSliceName, TState, AnySlice>({ ...opts, dependencies });
}
