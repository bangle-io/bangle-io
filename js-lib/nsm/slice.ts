import type {
  AnyFn,
  EffectsBase,
  InferSlicesKey,
  SliceBase,
  SliceKeyBase,
  Transaction,
} from './common';
import { mapObjectValues } from './common';
import type { StoreState } from './state';

export class SliceKey<K extends string, SS extends object, DS extends Slice[]>
  implements SliceKeyBase<K, SS>
{
  constructor(public key: K, public initState: SS, public dependencies?: DS) {}
}

type ResolveStoreStateIfRegistered<
  SSB extends StoreState,
  K extends string,
> = SSB extends StoreState<infer SLs>
  ? K extends InferSlicesKey<SLs>
    ? SSB
    : never
  : never;

type DependenciesState<DS extends Slice[]> = {
  [K in keyof DS]: DS[K]['key']['initState'];
};

type DependenciesResolvedState<DS extends Slice[]> = {
  [K in keyof DS]: ReturnType<DS[K]['resolveState']>;
};

type RecordToValues<T extends Record<string, any>> = T extends Record<
  string,
  infer P
>
  ? P
  : never;

type SelectorFn<SS, DS extends Slice[], T> = (
  sliceState: SS,
  storeState: StoreState<DS>,
) => T;

type ResolvedSelectors<SE extends Record<string, SelectorFn<any, any, any>>> = {
  [K in keyof SE]: SE[K] extends AnyFn ? ReturnType<SE[K]> : never;
};

export type RawAction<P extends any[], SS, DS extends Slice[]> = (
  ...payload: P
) => (sliceState: SS, storeState: StoreState<DS>) => SS;

export type Action<K extends string, P extends any[] = unknown[]> = (
  ...payload: P
) => Transaction<K, P>;

export class Slice<
  K extends string = string,
  SS extends object = any,
  DS extends Slice[] = any,
  A extends Record<string, RawAction<any[], SS, DS>> = any,
  SE extends Record<string, SelectorFn<SS, DS, any>> = any,
> implements SliceBase<K, SS>
{
  static create<
    K extends string,
    SS extends object,
    DS extends Slice[],
    A extends Record<string, RawAction<any[], SS, DS>>,
    SE extends Record<string, SelectorFn<SS, DS, any>>,
  >({
    key,
    initState,
    dependencies,
    actions,
    effects = [],
    selectors,
  }: {
    key: K;
    initState: SS;
    selectors?: SE;
    dependencies?: DS;
    actions?: A;
    effects?: EffectsBase[];
  }) {
    return new Slice(
      new SliceKey(key, initState, dependencies),
      actions,
      selectors,
    );
  }

  fingerPrint: string;
  actions: RawActionsToActions<K, A>;

  constructor(
    public key: SliceKey<K, SS, DS>,
    private _rawActions: A = {} as A,
    public selectors: SE = {} as SE,
  ) {
    this.fingerPrint = `${key.key}(${(key.dependencies || [])
      .map((d) => d.fingerPrint)
      .join(',')})`;
    this.actions = parseRawActions(key.key, _rawActions);
  }

  // applyTransaction<SSB extends StoreState>(
  //   tx: Transaction<any, any>,
  //   storeState: StoreState,
  // ): SS {
  //   const action: undefined | RawAction<any, SS, DS> =
  //     this._rawActions[tx.actionId];

  //   if (!action) {
  //     throw new Error(
  //       `Action "${tx.actionId}" not found in Slice "${this.key.key}"`,
  //     );
  //   }

  //   const sliceState = this.getState(storeState);
  //   const depState = this.getDependenciesState(storeState);
  //   const newState = action(...tx.payload)(sliceState, depState);

  //   return newState;
  // }

  // getDependenciesState<SSB extends StoreState>(
  //   storeState: ResolveStoreStateIfRegistered<SSB, K>,
  // ): DependenciesState<DS> {
  //   // ensure the slice is registered in the store
  //   this.getState(storeState);

  //   const result = (this.key.dependencies || []).map((slice) => {
  //     return slice.getState(storeState);
  //   });

  //   return result as DependenciesState<DS>;
  // }

  getState<SSB extends StoreState>(
    storeState: ResolveStoreStateIfRegistered<SSB, K>,
  ): SS {
    const result = storeState.getSliceState(this as SliceBase<any, any>);

    return result;
  }

  resolveSelectors<SSB extends StoreState>(
    storeState: ResolveStoreStateIfRegistered<SSB, K>,
  ): ResolvedSelectors<SE> {
    const result = mapObjectValues(this.selectors, (selector) => {
      return selector(this.getState(storeState), storeState);
    });

    return result as any;
  }

  // Returns the slice state with the selectors resolved
  resolveState<SSB extends StoreState>(
    storeState: ResolveStoreStateIfRegistered<SSB, K>,
  ): SS & ResolvedSelectors<SE> {
    return {
      ...this.getState(storeState),
      ...this.resolveSelectors(storeState),
    };
  }

  _getRawAction(actionId: string): RawAction<any, any, any> | undefined {
    return this._rawActions[actionId];
  }
}

export type RawActionsToActions<
  K extends string,
  A extends Record<string, RawAction<any, any, any>>,
> = {
  [KK in keyof A]: A[KK] extends (...param: infer P) => any
    ? Action<K, P>
    : never;
};

export function parseRawActions<
  K extends string,
  A extends Record<string, RawAction<any, any, any>>,
>(key: K, actions: A): RawActionsToActions<K, A> {
  let result = mapObjectValues(
    actions,
    (action, actionName): Action<K, any> => {
      return (...payload) => {
        return {
          sliceKey: key,
          actionId: actionName,
          payload,
        };
      };
    },
  );

  return result as RawActionsToActions<K, A>;
}
