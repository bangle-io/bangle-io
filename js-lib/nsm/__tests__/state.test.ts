import { expectType } from '../common';
import { Slice } from '../slice';
import { overrideInitState, StoreState } from '../state';

test('applyTransaction works', () => {
  const slice = Slice.create({
    key: 'test',
    initState: { num: 1 },
    actions: {
      myAction: (num: number) => (state) => {
        return { ...state, num: num + state.num };
      },
      action2: (num: number, foo: string, brother: () => void) => (state) =>
        state,
    },
  });

  const state = StoreState.create({
    storeName: 'test',
    slices: [slice],
  });

  expectType<number[]>(slice.actions.myAction(5).payload);
  expectType<[number, string, () => void]>(
    slice.actions.action2(5, 'str', () => {}).payload,
  );

  expect(slice.actions.myAction(5).payload).toEqual([5]);
  expect(slice.actions.action2(5, 'str', () => {}).payload).toEqual([
    5,
    'str',
    expect.any(Function),
  ]);

  const newState = state.applyTransaction(slice.actions.myAction(5))!;

  expect(newState.getSliceState(slice)).toEqual({
    num: 6,
  });
});

test('applyTransaction works with dependencies', () => {
  const sliceDep1 = Slice.create({
    key: 'test-dep1',
    initState: { num: 50 },
    actions: {},
  });

  const sliceDep2 = Slice.create({
    key: 'test-dep2',
    initState: { num: 3 },
    actions: {},
  });

  const slice = Slice.create({
    key: 'test',
    initState: { num: 1 },
    dependencies: { sliceDep1, sliceDep2 },
    actions: {
      myAction:
        (num: number) =>
        (state, { sliceDep1, sliceDep2 }) => {
          return {
            ...state,
            num: num + state.num + sliceDep1.num + sliceDep2.num,
          };
        },
    },
  });

  const state = StoreState.create({
    storeName: 'test',
    slices: [sliceDep1, sliceDep2, slice],
  });

  const newState = state.applyTransaction(slice.actions.myAction(5))!;

  expect(newState.getSliceState(slice)).toEqual({
    num: 50 + 1 + 5 + 3,
  });
});

test('throws error if slice key not unique', () => {
  const slice = Slice.create({
    key: 'test',
    initState: { num: 1 },
    actions: {
      myAction: (num: number) => (state) => {
        return { ...state, num: num + state.num };
      },
    },
  });

  expect(() => {
    StoreState.create({
      storeName: 'test',
      slices: [slice, slice],
    });
  }).toThrowError('Duplicate slice keys');
});

test('throws error if slice dependency is not registered', () => {
  const sliceDep = Slice.create({
    key: 'test-dep',
    initState: { num: 1 },
    actions: {},
  });
  const slice = Slice.create({
    key: 'test',
    initState: { num: 1 },
    dependencies: { sliceDep },
    actions: {
      myAction: (num: number) => (state) => {
        return { ...state, num: num + state.num };
      },
    },
  });

  expect(() => {
    StoreState.create({
      storeName: 'test',
      slices: [slice],
    });
  }).toThrowErrorMatchingInlineSnapshot(
    `"Slice "test" has a dependency on Slice "test-dep" which is either not registered or is registered after this slice."`,
  );
});

test('throws error if slice dependency is not registered before', () => {
  const sliceDep = Slice.create({
    key: 'test-dep',
    initState: { num: 1 },
    actions: {},
  });
  const slice = Slice.create({
    key: 'test',
    initState: { num: 1 },
    dependencies: { sliceDep },
    actions: {
      myAction: (num: number) => (state) => {
        return { ...state, num: num + state.num };
      },
    },
  });

  expect(() => {
    StoreState.create({
      storeName: 'test',
      slices: [slice, sliceDep],
    });
  }).toThrowErrorMatchingInlineSnapshot(
    `"Slice "test" has a dependency on Slice "test-dep" which is either not registered or is registered after this slice."`,
  );
});

test('overrideInitState works', () => {
  const slice1 = Slice.create({
    key: 'test1',
    initState: { num: 1 },
    actions: {},
  });

  const slice2 = Slice.create({
    key: 'test2',
    initState: { num: 2 },
    actions: {},
  });

  let newState = StoreState.create({
    storeName: 'test',
    slices: [slice1, overrideInitState(slice2, { num: 99 })],
  });

  expect(newState.getSliceState(slice1)).toEqual({ num: 1 });
  expect(newState.getSliceState(slice2)).toEqual({ num: 99 });

  newState = StoreState.create({
    storeName: 'test',
    slices: [overrideInitState(slice1, { num: -1 }), slice2],
  });

  expect(newState.getSliceState(slice1)).toEqual({ num: -1 });
});
