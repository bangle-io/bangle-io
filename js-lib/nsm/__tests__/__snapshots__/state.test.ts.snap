// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State creation with a slice 1`] = `
StoreState {
  "_slices": [
    Slice {
      "_actionSerializer": ActionSerializer {
        "getRawAction": [Function],
        "key": SliceKey {
          "dependencies": [],
          "initState": {
            "val": null,
          },
          "key": "mySlice",
          "selectors": {},
        },
        "rawActions": {},
      },
      "_flatDependencies": Set {},
      "_rawActions": {},
      "actions": {},
      "config": {},
      "effects": [],
      "fingerPrint": "mySlice()",
      "key": SliceKey {
        "dependencies": [],
        "initState": {
          "val": null,
        },
        "key": "mySlice",
        "selectors": {},
      },
    },
  ],
  "_transaction": undefined,
  "opts": undefined,
  "slicesCurrentState": {
    "mySlice": {
      "val": null,
    },
  },
}
`;
